File: vi/codes/python/modules/tree_node.py
"""
File: tree_node.py
Created Time: 2021-12-11
Author: krahets (krahets@163.com)
"""

from collections import deque


class TreeNode:
    """Lớp nút của cây nhị phân"""

    def __init__(self, val: int = 0):
        self.val: int = val  # Giá trị của nút
        self.height: int = 0  # Chiều cao của nút
        self.left: TreeNode | None = None  # Tham chiếu đến nút con trái
        self.right: TreeNode | None = None  # Tham chiếu đến nút con phải

    # Về quy tắc mã hóa tuần tự hóa, vui lòng tham khảo:
    # https://www.hello-algo.com/chapter_tree/array_representation_of_tree/
    # Biểu diễn mảng của cây nhị phân:
    # [1, 2, 3, 4, None, 6, 7, 8, 9, None, None, 12, None, None, 15]
    # Biểu diễn danh sách liên kết của cây nhị phân:
    #             /——— 15
    #         /——— 7
    #     /——— 3
    #    |    \——— 6
    #    |        \——— 12
    # ——— 1
    #     \——— 2
    #        |    /——— 9
    #         \——— 4
    #             \——— 8


def list_to_tree_dfs(arr: list[int], i: int) -> TreeNode | None:
    """Chuyển đổi một danh sách thành cây nhị phân: đệ quy"""
    # Nếu chỉ số vượt quá độ dài mảng, hoặc phần tử tương ứng là None, trả về None
    if i < 0 or i >= len(arr) or arr[i] is None:
        return None
    # Xây dựng nút hiện tại
    root = TreeNode(arr[i])
    # Đệ quy xây dựng cây con trái và cây con phải
    root.left = list_to_tree_dfs(arr, 2 * i + 1)
    root.right = list_to_tree_dfs(arr, 2 * i + 2)
    return root


def list_to_tree(arr: list[int]) -> TreeNode | None:
    """Chuyển đổi một danh sách thành cây nhị phân"""
    return list_to_tree_dfs(arr, 0)


def tree_to_list_dfs(root: TreeNode, i: int, res: list[int]) -> list[int]:
    """Chuyển đổi cây nhị phân thành danh sách: đệ quy"""
    if root is None:
        return
    if i >= len(res):
        res += [None] * (i - len(res) + 1)
    res[i] = root.val
    tree_to_list_dfs(root.left, 2 * i + 1, res)
    tree_to_list_dfs(root.right, 2 * i + 2, res)


def tree_to_list(root: TreeNode | None) -> list[int]:
    """Chuyển đổi cây nhị phân thành danh sách"""
    res = []
    tree_to_list_dfs(root, 0, res)
    return res

File: vi/codes/kotlin/utils/TreeNode.kt
/**
 * File: TreeNode.kt
 * Created Time: 2024-01-25
 * Author: curtishd (1023632660@qq.com)
 */

package utils

/* Lớp nút cây nhị phân */
/* Hàm khởi tạo */
class TreeNode(
    var _val: Int // Giá trị nút
) {
    var height: Int = 0 // Chiều cao nút
    var left: TreeNode? = null // Tham chiếu đến nút con trái
    var right: TreeNode? = null // Tham chiếu đến nút con phải

    // Về quy tắc mã hóa tuần tự hóa, vui lòng tham khảo:
    // https://www.hello-algo.com/chapter_tree/array_representation_of_tree/
    // Biểu diễn mảng của cây nhị phân:
    // [1, 2, 3, 4, None, 6, 7, 8, 9, None, None, 12, None, None, 15]
    // Biểu diễn danh sách liên kết của cây nhị phân:
    //             /——— 15
    //         /——— 7
    //     /——— 3
    //    |    \——— 6
    //    |        \——— 12
    // ——— 1
    //     \——— 2
    //        |    /——— 9
    //         \——— 4
    //             \——— 8

    /* Chuyển đổi một danh sách (list) thành cây nhị phân: đệ quy */
    companion object {
        private fun listToTreeDFS(arr: MutableList<Int?>, i: Int): TreeNode? {
            if (i < 0 || i >= arr.size || arr[i] == null) {
                return null
            }
            val root = TreeNode(arr[i]!!)
            root.left = listToTreeDFS(arr, 2 * i + 1)
            root.right = listToTreeDFS(arr, 2 * i + 2)
            return root
        }

        /* Chuyển đổi một danh sách (list) thành cây nhị phân */
        fun listToTree(arr: MutableList<Int?>): TreeNode? {
            return listToTreeDFS(arr, 0)
        }

        /* Chuyển đổi cây nhị phân thành danh sách (list): đệ quy */
        private fun treeToListDFS(root: TreeNode?, i: Int, res: MutableList<Int?>) {
            if (root == null) return
            while (i >= res.size) {
                res.add(null)
            }
            res[i] = root._val
            treeToListDFS(root.left, 2 * i + 1, res)
            treeToListDFS(root.right, 2 * i + 2, res)
        }

        /* Chuyển đổi cây nhị phân thành danh sách (list) */
        fun treeToList(root: TreeNode?): MutableList<Int?> {
            val res = mutableListOf<Int?>()
            treeToListDFS(root, 0, res)
            return res
        }
    }
}